test_py/test_make_request.py

```diff
@@ -0,0 +1,219 @@
+import json
+
+import pytest
+import requests
+from src_py.grndwork_api_client.make_request import (
+    ContentRange,
+    make_paginated_request,
+    make_request,
+    parse_content_range,
+)
+
+
+API_URL = 'https://api.grndwork.com/v1/tokens'
+
+
+@pytest.fixture(name='requests', autouse=True)
+def fixture_requests(mocker):
+    req_mock = mocker.patch(
+        target='src_py.grndwork_api_client.make_request.requests',
+        spec=requests,
+    )
+    response = mocker.MagicMock()
+    response.status_code = 200
+    response.json.return_value = {}
+    response.headers = {}
+    req_mock.request.return_value = response
+
+    return req_mock
+
+
+def describe_make_paginated_request():
+
+    @pytest.fixture(name='make_request', autouse=False)
+    def fixture_make_requests(mocker):
+        return mocker.patch(
+            target='src_py.grndwork_api_client.make_request.make_request',
+            spec=make_request,
+        )
+
+    def it_updates_offset_and_limit(make_request):
+        return_vals = [
+            (
+                ['' for i in range(20)],
+                {'Content-Range': 'items 6-25/65'},
+            ),
+            (
+                ['' for i in range(20)],
+                {'Content-Range': 'items 26-45/65'},
+            ),
+            (
+                ['' for i in range(20)],
+                {'Content-Range': 'items 46-65/65'},
+            ),
+        ]
+
+        make_request.side_effect = return_vals
+
+        pag_req = make_paginated_request(
+            url='url',
+            token='token',
+            page_size=20,
+            query={
+                'offset': 5,
+            },
+        )
+
+        assert(len(list(pag_req))) == 60
+
+        (_, kwargs) = make_request.call_args
+        assert kwargs['query']['limit'] == 20
+        assert kwargs['query']['offset'] == 45
+
+    def it_returns_when_last_record_reached(make_request):
+        return_vals = [
+            (
+                ['' for i in range(20)],
+                {'Content-Range': 'items 6-25/25'},
+            ),
+        ]
+
+        make_request.side_effect = return_vals
+
+        pag_req = make_paginated_request(
+            url='url',
+            token='token',
+            page_size=20,
+            query={
+                'offset': 5,
+            },
+        )
+
+        assert len(list(pag_req)) == 20
+
+    def it_returns_when_limit_reached(make_request):
+        return_vals = [
+            (
+                ['' for i in range(20)],
+                {'Content-Range': 'items 1-20/105'},
+            ),
+            (
+                ['' for i in range(20)],
+                {'Content-Range': 'items 21-40/105'},
+            ),
+            (
+                [''],
+                {'Content-Range': 'items 41-42/105'},
```

Comment:
```suggestion
                {'Content-Range': 'items 41-41/105'},
```

Reply:
I actually did that on purpose to make sure we're exercising this line: `if limit <= 0:`
Might be a cleaner way to do it though.

Reply:
Done